stan/math/rev/fun/quad_form.hpp

```diff
@@ -157,6 +216,49 @@ inline var quad_form(const EigMat& A, const ColVec& B, bool symmetric = false) {
   return baseVari->impl_->C_(0, 0);
 }
 
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * Symmetry of the resulting matrix is not guaranteed due to numerical
+ * precision.
+ *
+ * @tparam EigMat1 type of the first (square) matrix
+ * @tparam EigMat2 type of the second matrix
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @return The quadratic form, which is a symmetric matrix.
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_not_col_vector_t<Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline auto quad_form(const Mat1& A, const Mat2& B, bool symmetric = false) {
+  return internal::quad_form_impl(A, B, symmetric);
+}
+
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * @tparam EigMat type of the matrix
+ * @tparam ColVec type of the vector
+ *
+ * @param A square matrix
+ * @param B vector
+ * @return The quadratic form (a scalar).
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat, typename Vec,
+	  require_matrix_t<Mat>* = nullptr,
+          require_col_vector_t<Vec>* = nullptr,
+          require_any_var_matrix_t<Mat, Vec>* = nullptr>
+inline var quad_form(const Mat& A, const Vec& B, bool symmetric = false) {
+  return internal::quad_form_impl(A, B, symmetric)(0, 0);
```

Comment:
This access here creates an extra var and an extra access. I can do more verbose code and avoid this if necessary.

Reply:
I think this is not necessary. Eventually we will have indexing not create extra vars.